x-pack/plugins/endpoint/common/types.ts

```diff
@@ -351,59 +351,67 @@ export type PageId = 'alertsPage' | 'managementPage' | 'policyListPage';
  */
-type KbnConfigSchemaInputTypeOf<
-  T extends kbnConfigSchemaTypes.Type<unknown>
-> = T extends kbnConfigSchemaTypes.ObjectType
+type KbnConfigSchemaInputTypeOf<T> = T extends Record<string, unknown>
   ? KbnConfigSchemaInputObjectTypeOf<
       T
     > /** `schema.number()` accepts strings, so this type should accept them as well. */
-  : kbnConfigSchemaTypes.Type<number> extends T
-  ? TypeOf<T> | string
-  : TypeOf<T>;
+  : number extends T
+  ? T | string
+  : T;
 
 /**
  * Works like ObjectResultType, except that 'maybe' schema will create an optional key.
  * This allows us to avoid passing 'maybeKey: undefined' when constructing such an object.
  *
  * Instead of using this directly, use `InputTypeOf`.
  */
-type KbnConfigSchemaInputObjectTypeOf<
-  T extends kbnConfigSchemaTypes.ObjectType
-> = T extends kbnConfigSchemaTypes.ObjectType<infer P>
-  ? {
-      /** Use ? to make the field optional if the prop accepts undefined.
-       * This allows us to avoid writing `field: undefined` for optional fields.
-       */
-      [K in Exclude<
-        keyof P,
-        keyof KbnConfigSchemaNonOptionalProps<P>
-      >]?: KbnConfigSchemaInputTypeOf<P[K]>;
-    } &
-      { [K in keyof KbnConfigSchemaNonOptionalProps<P>]: KbnConfigSchemaInputTypeOf<P[K]> }
-  : never;
+type KbnConfigSchemaInputObjectTypeOf<P extends Record<string, unknown>> = {
+  /** Use ? to make the field optional if the prop accepts undefined.
+   * This allows us to avoid writing `field: undefined` for optional fields.
+   */
+  [K in Exclude<keyof P, keyof KbnConfigSchemaNonOptionalProps<P>>]?: KbnConfigSchemaInputTypeOf<
+    P[K]
+  >;
+} &
+  { [K in keyof KbnConfigSchemaNonOptionalProps<P>]: KbnConfigSchemaInputTypeOf<P[K]> };
 
 /**
  * Takes the props of a schema.object type, and returns a version that excludes
  * optional values. Used by `InputObjectTypeOf`.
  *
  * Instead of using this directly, use `InputTypeOf`.
  */
-type KbnConfigSchemaNonOptionalProps<Props extends kbnConfigSchemaTypes.Props> = Pick<
+type KbnConfigSchemaNonOptionalProps<Props extends Record<string, unknown>> = Pick<
   Props,
   {
-    [Key in keyof Props]: undefined extends TypeOf<Props[Key]> ? never : Key;
+    [Key in keyof Props]: undefined extends Props[Key]
+      ? never
+      : null extends Props[Key]
+      ? never
+      : Key;
   }[keyof Props]
 >;
 
 /**
  * Query params to pass to the alert API when fetching new data.
  */
 export type AlertingIndexGetQueryInput = KbnConfigSchemaInputTypeOf<
-  typeof alertingIndexGetQuerySchema
+  TypeOf<typeof alertingIndexGetQuerySchema>
 >;
 
 /**
  * Result of the validated query params when handling alert index requests.
  */
 export type AlertingIndexGetQueryResult = TypeOf<typeof alertingIndexGetQuerySchema>;
+
+/**
+ * Request body for the 'metadata' API index
+ */
+export type MetadataIndexGetBodyInput = KbnConfigSchemaInputTypeOf<
+  TypeOf<typeof metadataIndexGetBodySchema>
+>;
```

Comment:
Why do we need all the generics. Why not a simple schema defined as and when needed for the particular use case, like we did for the configuration. I may be wrong, but this looks some future proofing. My apprehension is that this couples the whole system (alert, endpoint) etc together. I think we should keep things simple and separate.

Reply:
@nnamdifrankie I don't understand. The reason for the change is documented here: https://github.com/elastic/endpoint-app-team/issues/177

Is your concern with the implementation of the idea, or is your concern w/ the idea?

Reply:
I am fine with the idea and I think it is a good idea, like how you will have with swagger, grapql etc. And I see them do it in fleet. I am question the generic wrappers.

Reply:
Gotcha. The types and patterns used here were introduced in this PR: https://github.com/elastic/kibana/pull/59340
I should've included this info in the PR but forgot.

To summarize and clarify:`@kbn/config-schema` schemas are used to validate the inputs to HTTP handlers. The `@kbn/config-schema` package provides a generic type alias,`TypeOf`, which is used to calculate the type returned by the schema's `validate` method. HTTP handlers receive query params, request bodies, and route params which are typed this way. This pattern is existing in Kibana and in our plugin and isn't changed by this PR. 

When creating a request to one of our HTTP handlers in the client, I want the query params or request body to be typed based on the same schema. The type called `KbnConfigSchemaInputTypeOf` is used to do this. It is similar to the provided `TypeOf` type alias however instead of representing the result of the validation, it represents the expected input to the schema. If our team uses this pattern successfully, I will petition the team that maintains `@kbn/config-schema` to include such a type in the official package. Other teams are interested pursuing similar approaches.

With all of that in mind, do you feel that this approach is good? With this approach a developer can create a schema using `@kbn/config-schema` and then calculate a type that can be used by the client. The schema will also be used to calculate the type passed to the http handler.